File: C++/061. Rotate List.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (k == 0 || head == nullptr)
            return head;
        ListNode *p, *q;
        int n;
        for (n = 0, p = head; p != nullptr; n++, p = p->next);
        k = k % n;
        ListNode *res = new ListNode(0);
        res->next = head;
        p = res, q = res;
        for (int i = 0; i < k; i++, q = q->next);
        for (; q->next != nullptr; p = p->next, q = q->next);
        q->next = res->next;
        res->next = p->next;
        p->next = nullptr;
        return res->next;
    }
};

File: Go/061. Rotate List.go
package main

type ListNode struct {
    Val int
    Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || k == 0 {
		return head
	}
	var p, q *ListNode
	var n int
	for p, n = head, 0; p != nil; p, n = p.Next, n + 1 {}
	k = k % n
	res := &ListNode{0, head}
	p, q = res, res
	for i := 0; i < k; i, q = i + 1, q.Next {}
	for ; q.Next != nil; p, q = p.Next, q.Next {}
	q.Next = res.Next
	res.Next = p.Next
	p.Next = nil
	return res.Next
}